wallet/src/account/output_cache/mod.rs

```diff
@@ -890,36 +890,33 @@ impl OutputCache {
     }
 
     fn violates_frozen_token(&self, unconfirmed_tx: &WalletTx, frozen_token_id: &TokenId) -> bool {
-        // We check only inputs because currently it's not possible to have a token in an output
-        // without having it in an input.
-        // But potentially we should check outputs too.
-        unconfirmed_tx.inputs().iter().any(|inp| match inp {
+        let output_uses_frozen_token = |output: &TxOutput| match output {
+            TxOutput::Transfer(v, _)
+            | TxOutput::LockThenTransfer(v, _, _)
+            | TxOutput::Burn(v)
+            | TxOutput::Htlc(v, _) => match v {
+                OutputValue::TokenV1(token_id, _) => frozen_token_id == token_id,
+                OutputValue::TokenV0(_) | OutputValue::Coin(_) => false,
+            },
+            TxOutput::CreateOrder(data) => [data.ask(), data.give()].iter().any(|v| match v {
+                OutputValue::TokenV1(token_id, _) => frozen_token_id == token_id,
+                OutputValue::TokenV0(_) | OutputValue::Coin(_) => false,
+            }),
+            TxOutput::IssueNft(_, _, _)
+            | TxOutput::DataDeposit(_)
+            | TxOutput::CreateStakePool(_, _)
+            | TxOutput::DelegateStaking(_, _)
+            | TxOutput::CreateDelegationId(_, _)
+            | TxOutput::IssueFungibleToken(_)
+            | TxOutput::ProduceBlockFromStake(_, _) => false,
+        };
```

Comment:
maybe we can reuse get_referenced_token_ids from token_utils, something like:
get_referenced_token_ids(output).contains(frozen_token_id)

Reply:
> Or maybe even change get_referenced_token_ids to return an iterator instead of the BTreeSet and let the caller collect if needed.

In the end I decided to return a `SmallVec<[TokenId; 2]>` instead (returning `impl Iterator` is also possible, but it looked less clean IMO).